x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DeprecationRoleDescriptorConsumer.java

```diff
@@ -0,0 +1,211 @@
+
+    private static final String ROLE_PERMISSION_DEPRECATION_STANZA = "Role [%s] contains index privileges covering the [%s] alias but"
+            + " which do not cover some of the indices that it points to [%s]. Granting privileges over an alias and hence granting"
+            + " privileges over all the indices that the alias points to is deprecated and will be removed in a future version of"
+            + " Elasticsearch. Instead define permissions exclusively on index names or index name patterns.";
+
+    private static final Logger logger = LogManager.getLogger(DeprecationRoleDescriptorConsumer.class);
+
+    private final DeprecationLogger deprecationLogger;
+    private final ClusterService clusterService;
+    private final ThreadPool threadPool;
+    private final Object mutex;
+    private final Queue<RoleDescriptor> workQueue;
+    private boolean workerBusy;
+    private final Set<String> dailyRoleCache;
+
+    public DeprecationRoleDescriptorConsumer(ClusterService clusterService, ThreadPool threadPool) {
+        this(clusterService, threadPool, new DeprecationLogger(logger));
+    }
+
+    // package-private for testing
+    DeprecationRoleDescriptorConsumer(ClusterService clusterService, ThreadPool threadPool, DeprecationLogger deprecationLogger) {
+        this.deprecationLogger = deprecationLogger;
+        this.clusterService = clusterService;
+        this.threadPool = threadPool;
+        this.mutex = new Object();
+        this.workQueue = new LinkedList<>();
+        this.workerBusy = false;
+        // this String Set keeps "<date>-<role>" pairs so that we only log a role once a day.
+        this.dailyRoleCache = Collections.newSetFromMap(new LinkedHashMap<String, Boolean>() {
+            @Override
+            protected boolean removeEldestEntry(final Map.Entry<String, Boolean> eldest) {
+                return false == eldest.getKey().startsWith(todayISODate());
+            }
+        });
+    }
+
+    @Override
+    public void accept(Collection<RoleDescriptor> effectiveRoleDescriptors) {
+        synchronized (mutex) {
+            for (RoleDescriptor roleDescriptor : effectiveRoleDescriptors) {
+                if (dailyRoleCache.add(buildCacheKey(roleDescriptor))) {
+                    workQueue.add(roleDescriptor);
+                }
+            }
+            if (false == workerBusy) {
+                workerBusy = true;
+                try {
+                    // spawn another worker on the generic thread pool
+                    threadPool.generic().execute(new AbstractRunnable() {
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            logger.warn("Failed to produce role deprecation messages", e);
+                            synchronized (mutex) {
+                                workerBusy = false;
+                            }
```

Comment:
I am on the fence about what to do in this failure case. The current way is to drop the offending role descriptor and stop the worker. But we could conserve the offending role descriptor, or retry, or, on the contrary, blacklist it. Moreover, we could not stop the worker but let it continue, although this brings little difference compared to this proposal.

Given that such exception are rare, and the consequences of skipping a `RoleDescriptor` work item are of little relevance, I believe that skipping it for now, and retrying the next day is fine. Blacklisting the role, is a close competing alternative - I am thinking of the `TooComplexToDeterminize` exception - but getting this exception in the logs indicates a complex role which is a problem in itself that we might mask by blacklisting.

Also, I might overthink it, so I'll consider any of your suggestions.

Reply:
I think dropping the bad descriptor is the right thing to do; blacklisting probably isn't necessary. Maybe we could just keep working if there is more work?


```java
synchronized (mutex) {
    final boolean hasMoreWork = workQueue.peek() != null;
    if (hasMoreWork) {
        workerBusy = true; // just being paranoid :)
        threadPool.generic().execute(this);
    } else {
        workerBusy = false;
    }
}
```